### DISCUSSION TITLE
ConfigurationLoadCredential failed

### DISCUSSION THREAD

**Discussion by yc1997427** [microsoft/msquic#1137] [Q&A] [answered]
Hi, I tried to run sample code to run server, with command `./quicsample -server -cert_file:~/quic/msquic/certs/out/leaf_cert.pem -key_file:~/quic/msquic/certs/out/leaf_cert.pkcs8`.

But kept getting error ConfigurationLoadCredential failed, 0xbebc212!
leaf_cert.pem and  leaf_cert.pkcs8 are files I generated from google generate_certficate code.

**Comment by ThadHouse** [comment]
Just to verify, you're running linux, right? And do you have the exact code you used to generate a certificate? That way we can try reproducing the issue. Its possible its an issue with the certificate.

**Reply by yc1997427** [reply]
Yes, I'm running on Ubuntu 18.04. I used https://chromium.googlesource.com/chromium/src/+/master/net/tools/quic/certs/generate-certs.sh. The google script to generate certs and key.

**Reply by yc1997427** [reply]
I think it would be very helpful if msquic can provide example cert and key which can be used immediately. Quiche has these, which makes it much easier to run the tests.

**Reply by ThadHouse** [reply]
I agree. I'll look into this to see if we can make that experience better.

**Reply by ThadHouse** [reply]
So the reason that script doesn't work is currently msquic does not support certificates with a pass phrase, and that script won't let me generate a key without one. 

The following openssl command was able to create a certificate that was usable from msquic, and I could connect to.

**Comment by okbaswc** [comment]
What exactly is the right way to generate the certificates on Windows for these two files? I've tried the Google script mentioned above as well as doing it the `New-SelfSignedCertificate` way and exporting/extracting cert and private key with OpenSSL, and nothing seems to work. On Windows, what 2 files are expected as input here?

**Reply by ThadHouse** [reply]
For windows schannel, using files is not supported, and the certificate store must be used. For windows openssl, we added support recently to use the cert store as well. The only requirement is that when the cert is created or imported, its marked as exportable. Then you just can pass in the thumbprint. That is our recommended way. Otherwise, I believe that you would need to generate Linux style certs to pass into openssl as files. There likely are ways to convert a windows cert to a Linux openssl cert, but I don't have any recommendations.

**Comment by ThadHouse** [comment, marked as answer]
#1151 adds some documentation on the sample on how to create self signed certificates for both Linux and Windows, for OpenSSL and Schannel providers. Hopefully that helps with some of these issues, and allows creating certificates that are known to work.

**Reply by yc1997427** [reply]
Hi, it seems the cert and key are still not working. 
I run ./quicsample -server -cert_file:~/quic/msquic/server.cert -key_file:~/quic/msquic/server.key
and got, ConfigurationLoadCredential failed, 0xbebc212!

**Reply by ThadHouse** [reply]
Can you post your binaries and the cert you're trying to use in a Zip attached to this? That way I can do some double checking to see if there is another issue.

**Reply by ThadHouse** [reply]
Hmm. Finally figured out what gives the error. 0xbebc212 happens when openssl can't find the certificate files. Are you sure thats the right directory, and that theyre not empty files?

**Reply by yc1997427** [reply]
Really appreciate for checking. Yes, thats the right directory and files are not empty. 
This is the printout of what the directory has. 
~/quic/msquic$ ls
artifacts       libicu55_55.1-7ubuntu0.5_amd64.deb        scripts
build           libssl1.0.0_1.0.2g-1ubuntu4.18_amd64.deb  server.cert
certs           LICENSE                                   server.key
cmake           msquic.kernel.sln                         src
CMakeLists.txt  nuget                                     submodules
Dockerfile      packages-microsoft-prod.deb
docs            README.md

And checked the key and server files are not emtpy. Also attached the zip file for cert and key. Thanks 
[cert.zip](https://github.com/microsoft/msquic/files/5810687/cert.zip)

**Reply by ThadHouse** [reply]
Ok, I think its actually the shell not handling `~` correctly. Those certs work if I use powershell with `~/quic/msquic/...`, but if I use bash with `~/quic/msquic/...` that does not and gives the error you're seeing. Does it work if you use a full path without `~`?

**Reply by yc1997427** [reply]
hi, I just tried without ~, it still throw out the same error...

**Reply by ThadHouse** [reply]
Ok, Powershell is actually expanding `~` before starting the application. So its passing to OpenSSL the full path. However, bash is not expanding it, and its likely OpenSSL internally doesn't do expansion when attempting to look up paths. I'll take a look, but there likely nothing we could do from our end, other then to document that full paths must be used.

**Reply by ThadHouse** [reply]
I just added the following lines at line 410 of the sample. If you add them and build, what extra prints? Just want to make sure it is getting a path that exists.

```
        printf("Cert File: %s\n", Cert);
        printf("Key File: %s\n", KeyFile);
```

**Reply by yc1997427** [reply]
Oh, it works now. I moved the files to the sample script directory, no need to put full path.

**Reply by ThadHouse** [reply]
Awesome. Glad its working. We just pass the paths directly to OpenSSL, which internally does all the file handling. I'm not sure if they have any documentation on any restrictions for that. But glad we got it working.

**Reply by yc1997427** [reply]
Just wondering what kind of value we need to put in the sample code to specify the IP address we want to set the server to. I tried to add "192.168.0.1" to address varible, but throwing out error. I'm intending to host the server into a network namespace, with the specifid address

**Comment by Wilson-Ian** [comment]
I add the info here about how to make it work.
https://github.com/microsoft/msquic/issues/1092#issuecomment-847733808
